src/treelearner/feature_histogram.hpp

```diff
@@ -211,13 +213,28 @@ class FeatureHistogram {
 
 private:
   /*!
+  * \brief Return the L1 regularized sum_gradients
+  * \param sum_gradients
+  * \return regularized sum_gradients
+  */
+  double L1RegularizedGradients(double sum_gradients) const {
+  if (sum_gradients > lambda_l1_) return sum_gradients - lambda_l1_;
+  if (sum_gradients < -lambda_l1_) return sum_gradients + lambda_l1_;
+  return 0.0f;
+  }
+
+  /*!
   * \brief Calculate the split gain based on sum_gradients and sum_hessians
   * \param sum_gradients
   * \param sum_hessians
   * \return split gain
   */
   double GetLeafSplitGain(double sum_gradients, double sum_hessians) const {
-    return (sum_gradients * sum_gradients) / (sum_hessians);
+    if (lambda_l1_ != 0.0f) {
```

Comment:
The member function in .h/.hpp is ```inline``` automatically. So the cost of calling function is very small. 
So we can move this into ```L1RegularizedGradients``` , : ```if ( lambda_l1_ == 0.0f) { return sum_gradients; }```
However, it still have 3 ```if```'s conditions at worst case, can we reduce this ?